neo4j-to-neptune/src/main/java/com/amazonaws/services/neptune/util/NeptuneBulkLoader.java

```diff
@@ -0,0 +1,397 @@
+    private final ObjectMapper objectMapper;
+
+    public NeptuneBulkLoader(
+            String bucketName,
+            String s3Prefix,
+            Region region,
+            String neptuneEndpoint,
+            String iamRoleArn) {
+
+        this.bucketName = bucketName;
+        this.s3Prefix = s3Prefix;
+        this.region = region;
+        this.neptuneEndpoint = neptuneEndpoint;
+        this.iamRoleArn = iamRoleArn;
+        this.objectMapper = new ObjectMapper();
+
+        this.s3AsyncClient = S3AsyncClient.builder()
+                .region(region)
+                .credentialsProvider(DefaultCredentialsProvider.create())
+                .build();
+
+        this.httpClient = HttpClient.newBuilder()
+                .connectTimeout(Duration.ofSeconds(120))
+                .build();
+
+        // Validate for not empty parameters
+        if (bucketName.trim().isEmpty()) {
+            throw new IllegalArgumentException("S3 bucket name cannot be empty");
+        }
+        if (s3Prefix.trim().isEmpty()) {
+            throw new IllegalArgumentException("S3 prefix cannot be empty");
+        }
+        if (region.toString().trim().isEmpty()) {
+            throw new IllegalArgumentException("Region cannot be empty");
+        }
+        if (neptuneEndpoint.trim().isEmpty()) {
+            throw new IllegalArgumentException("Neptune endpoint cannot be empty");
+        }
+        if (iamRoleArn.trim().isEmpty()) {
+            throw new IllegalArgumentException("IAM Role Arn cannot be empty");
+        }
+        System.out.println("S3 Bucket: " + bucketName);
+        System.out.println("S3 Prefix: " + s3Prefix);
+        System.out.println("AWS Region: " + region);
+        System.out.println("Neptune Endpoint: " + neptuneEndpoint);
+        System.out.println("IAM Role ARN: " + iamRoleArn);
+    }
+
+    /**
+     * Upload Neptune vertices and edges CSV files asynchronously
+     */
+    public void uploadCsvFilesToS3(String filePath) throws Exception {
+        System.out.println("Uploading Gremlin CSV to S3...");
+
+        // Grab the timestamp of ConvertCvs to use as S3 directory key prefix
+        String convertCsvTimeStamp = filePath.substring(filePath.lastIndexOf('/') + 1);
+
+        // Check if the S3 prefix is provided, and construct the full S3 key prefix using convertCsvTimeStamp
+        String s3KeyPrefixTimeStamp = Optional.ofNullable(s3Prefix)
+            .filter(prefix  -> !prefix.isEmpty())
+            .map(prefix  -> prefix + "/")
+            .orElse("") + convertCsvTimeStamp;
+
+        // Start both uploads concurrently
+        CompletableFuture<Boolean> verticesFuture = uploadFileAsync(
+            filePath + File.separator + "vertices.csv",
+            s3KeyPrefixTimeStamp + "/" + "vertices.csv"
+        );
+        CompletableFuture<Boolean> edgesFuture = uploadFileAsync(
+            filePath + File.separator + "edges.csv",
+            s3KeyPrefixTimeStamp + "/" + "edges.csv"
+        );
```

Comment:
Note: Eventually I think the plan is to make the csv conversion part of this utility multithreaded, which may result in multiples files for vertices and edges. Setting this up as a directory upload may be more future proofed.